i.MX RT700 DSP support

iuliana-prodan opened the pull request:
The i.MX RT700 has a Compute Subsystem which includes a primary ARM Cortex-M33 running at 325 MHz and Cadence Tensilica HiFi 4 DSP.
It also features an ultra-low power Sense Subsystem which includes a second ARM Cortex-M33 and Cadence Tensilica HiFi 1 DSP.

This pull request adds the soc and board support for the two HiFi cores.

iuliana-prodan committed:
soc: nxp: imxrt700: Add i.MXRT700 HiFi4 DSP support

The i.MX RT700 has a compute subsystem which includes
a primary ARM Cortex-M33 running at 325 MHz and
Cadence Tensilica HiFi4 DSP.

Here, we add support for the HiFi4 core.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
dts: xtensa: nxp: Add device tree for HiFi4 core from NXP i.MXRT700

Add a basic device tree for the HiFi4 DSP from NXP i.MXRT700.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
boards: nxp: imxrt700: Add i.MXRT700 HiFi4 board support

Add board support for the HiFi4 DSP core from NXP i.MXRT700.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
west.yml: Update Xtensa hal

Update Xtensa hal to support NXP i.MXRT700 DSP cores.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
soc: nxp: imxrt700: Add i.MXRT700 HiFi1 DSP support

The i.MX RT700 has an ultra-low power Sense Subsystem
which includes an ARM Cortex-M33 and
Cadence Tensilica HiFi 1 DSP.

Here, we add support for the HiFi1 core.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
dts: xtensa: nxp: Add device tree for HiFi1 core from NXP i.MXRT700

Add a basic device tree for the HiFi1 DSP from NXP i.MXRT700.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan committed:
boards: nxp: imxrt700: Add i.MXRT700 HiFi1 board support

Add board support for the HiFi1 DSP core from NXP i.MXRT700.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

dcpleung commented on soc/nxp/imxrt/imxrt7xx/Kconfig.defconfig:
```diff
@@ -0,0 +1,47 @@
+config SYS_CLOCK_TICKS_PER_SEC
+	default 1000
+
+config DYNAMIC_INTERRUPTS
+	default n
+
+config CACHE
+	default n
+
+config DCACHE
+	default n
+
+config CACHE_MANAGEMENT
+	default n
```
These are disabled by default so there is no need to set default to 'n'. Ditto for `SMP` and `*_LEVEL_INTERRUPTS` above.

iuliana-prodan commented:
The twister tests are failing because the CI is using zephyr_sdk 0.16.8 which doesn't include the hifi1 and hifi4 toolchain for imxrt700.
```
Configuration saved to '/__w/zephyr/zephyr/twister-out/mimxrt700_evk_mimxrt798s_hifi1/samples/kernel/condition_variables/simple/sample.kernel.cond_var.simple/zephyr/.config'
Kconfig header saved to '/__w/zephyr/zephyr/twister-out/mimxrt700_evk_mimxrt798s_hifi1/samples/kernel/condition_variables/simple/sample.kernel.cond_var.simple/zephyr/include/generated/zephyr/autoconf.h'
CMake Error at /__w/zephyr/zephyr/cmake/compiler/gcc/target.cmake:12 (message):
  C compiler
  /opt/toolchains/zephyr-sdk-0.16.8/xtensa-nxp_rt700_hifi1_zephyr-elf/bin/xtensa-nxp_rt700_hifi1_zephyr-elf-gcc
  not found - Please check your toolchain installation
  ```
  
  Locally I used zephyr-sdk-0.16.9-rc2 which had the 2 toolchain.
  
  I see now that 0.17.0 was released.
   @stephanosio when will the new toolchain be included in the CI?
  Can I do the update?

danieldegrasse commented on boards/nxp/mimxrt700_evk/board.cmake:
```diff
@@ -0,0 +1,16 @@
+
+if(CONFIG_SOC_MIMXRT798S_HIFI4)
+  board_runner_args(jlink "--device=MIMXRT798S_hifi4" "--reset-after-load")
+  board_runner_args(linkserver  "--device=MIMXRT798S:EVK-MIMXRT798")
```
For linkserver do we need the `--core` argument, since this isn't the primary core?

iuliana-prodan replied:
I haven't tested this yet and, since I'm not that familiar with i.MXRTs, I don't know exactly if this is needed or not. 
I can remove it, until I can test it. If needed, I'll add it later.

danieldegrasse commented on boards/nxp/mimxrt700_evk/Kconfig.mimxrt700_evk:
We should add a doc page here as well for the board, right? Otherwise, it won't be clear to users how to use this board support

iuliana-prodan replied:
I was thinking to add information about the DSPs in here: https://github.com/zephyrproject-rtos/zephyr/pull/79376/files#diff-4b9fba8cde414e167f42713372629679a50c662d6e9971da23ae3aef778e5c03, after @lucien-nxp's PR will be merged.
But, I can add one here and we merge the `index.rst` file when the last PR will be merged.

danieldegrasse replied:
Up to you. I thought this PR might merge before @lucien-nxp's did, but if not we can wait

iuliana-prodan replied:
I add an `index.rst` similar with the one from @lucien-nxp PR with info about the DSP cores.
As I already mentioned this will be merged with the one for Cortex M33 cores.

stephanosio commented:
> I see now that 0.17.0 was released.
> @stephanosio when will the new toolchain be included in the CI?
> Can I do the update?

I will send a PR for that in a few days after the CI image containing SDK 0.17.0 is ready and tested.

iuliana-prodan commented:
> > I see now that 0.17.0 was released.
> > @stephanosio when will the new toolchain be included in the CI?
> > Can I do the update?
> 
> I will send a PR for that in a few days after the CI image containing SDK 0.17.0 is ready and tested.

Great, thank you @stephanosio

iuliana-prodan committed:
boards: nxp: imxrt700: Add documentation for i.MXRT700 DSP cores

Add documentation for the two HiFi1 and HiFi4 DSP cores from i.MXRT700.

Signed-off-by: Iuliana Prodan <iuliana.prodan@nxp.com>

iuliana-prodan commented:
Changes since previous version:
- remove unused settings for debugging of DSP cores;
- add `index.rst` for documentation.

iuliana-prodan commented:
2 tests are failing in the CI
```
1) tests/drivers/spi/spi_loopback/drivers.spi.s32z_dspi.loopback on s32z2xxdc2@B/s32z270/rtu0 error (Cmake build failure)
2) tests/drivers/spi/spi_loopback/drivers.spi.s32z_dspi.loopback on s32z2xxdc2@B/s32z270/rtu1 error (Cmake build failure)
```
because in the [hal_nxp PR #444](https://github.com/zephyrproject-rtos/hal_nxp/pull/444) is missing the newly included commit - https://github.com/zephyrproject-rtos/hal_nxp/commit/683c007c3b834012358b4adf96a532d18ce05646

iuliana-prodan commented:
Changes since previous version:
- update `west.yml` with commit id for Xtensa HAL since the PR got merged;
- fixed all CI checks and tests.

dcpleung commented:
Waiting for hal_nxp to be updated.

dleach02 commented:
@iuliana-prodan the RT700 HAL PR merged. Can you update this PR to move the NXP HAL SHA to the tip and then resolve the merge conflicts? We can get this ready to merge quickly after the upstream v4.1 release

iuliana-prodan commented:
> @iuliana-prodan the RT700 HAL PR merged. Can you update this PR to move the NXP HAL SHA to the tip and then resolve the merge conflicts? We can get this ready to merge quickly after the upstream v4.1 release

Yes, I have this on my list :)
I'll do it this week!

iuliana-prodan commented:
Changes since previous version:
- rebase and resolve conflicts with the M33 Core 0 and Core1 support for i.MXRT700;
- update `west.yml` only with commit id for Xtensa HAL (NXP hal was merged and updated in `west.yml`);

iuliana-prodan commented on west.yml:
```diff
@@ -268,7 +268,7 @@ manifest:
         - hal
     - name: hal_xtensa
-      revision: baa56aa3e119b5aae43d16f9b2d2c8112e052871
+      revision: b38620c7cc61e349e192ed86a54940a5cd0636b7
```
right! I've added first.

iuliana-prodan commented on boards/nxp/mimxrt700_evk/mimxrt700_evk_mimxrt798s_hifi4_defconfig:
```diff
@@ -0,0 +1,10 @@
+CONFIG_GEN_ISR_TABLES=y
+CONFIG_GEN_IRQ_VECTOR_TABLE=n
+CONFIG_XTENSA_SMALL_VECTOR_TABLE_ENTRY=y
+CONFIG_NXP_IMXRT_BOOT_HEADER=n
```
moved them to soc (in `Kconfig.defconfig`) but some are with `default n` since they are defined with` default y` - see [CONFIG_GEN_IRQ_VECTOR_TABLE](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/Kconfig#L494) and [CONFIG_NXP_IMXRT_BOOT_HEADER](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nxp/imxrt/Kconfig#L24)

iuliana-prodan commented on soc/nxp/imxrt/imxrt7xx/hifi4/linker.ld:
```diff
@@ -0,0 +1,492 @@
+
+MEMORY
+{
+  vector_reset_text :
```
Most linker script files use spaces, with only a few (such as [mips](https://github.com/zephyrproject-rtos/zephyr/blob/91cf42c657afcc613f215f8af2f8e78fc8d0ffa2/include/zephyr/arch/mips/linker.ld)) using tabs. Therefore, I’ve kept the linker scripts for hifi1 and hifi4 formatted with spaces. 
If you strongly prefer tabs, I can update them, but in my opinion, this approach aligns with Zephyr’s current formatting.

nordicjm replied:
Seems there is a mix of space and tabs in different files, will allow

iuliana-prodan commented:
@nordicjm Thanks for reviewing.
I've resolved all you comments, except [this ](https://github.com/zephyrproject-rtos/zephyr/pull/79954#discussion_r1980825786)

dcpleung commented:
Removing the DNM tag since the Zephyr bot tagged it using an old tag.
